### PR TITLE
FEATURE: Link to text customization when editing system badges

### PR DESCRIPTION
**Pull request by tshenry**
Being that system badges ship with every instance of Discourse, we've opted to define the name, description, and long description in our locales files to promote translation into other languages. When an admin visits the overview page of a system badge in their admin panel, they are met with disabled inputs for these text properties. The problem is that we fail to educate the admin that the text needs to be managed via the site text customization settings. This can lead to support requests.

This change will add a "Customize text" link under the disabled inputs that will take the user to the relevant site text customization.

Screen Shot 2020-11-24 at 5 50 52 PM

In the future, it may be nice to enable the inputs, connecting them directly to the text customization, but this seemed like a quick and easy way to improve the current situation.

Noting that slugs use dashes and the text customization keys use underscores when badge names include multiple words.  The `text_customization_name` property handles this conversion.

This seems like a simple change, but if there is a specific test that would make sense for this, please let me know.

### CODE REVIEW

**Review by davidtaylorhq — COMMENTED** [review]
This looks good - I like the UI and I agree this is a good simple solution to the problem.

My only suggestion is that we should try to avoid having all that HTML in the translation string. It would be better to only have the text in there. We can also use the `link-to` helper to automagically build the URL, without needing to worry about `basePath` or anything like that. So the handlebars would look a little like: (untested)

```
{{#link-to "adminSiteText.index" (query-params q=(concat 'badges.' text_customization_name ".long_description")}}
  {{i18n 'mylabel'}}
{{/link-to}}
```
Bonus points for refactoring the `concat` logic into the js controller, rather than handlebars. But I'm not sure what the best way would be to do that, given that the last segment is different each time. Maybe go halfway and make a `text_customization_prefix` computed variable which is `badges.{name}.`, so the concat is just `(concat text_customization_prefix "long_description")`

**Comment by tshenry** [comment]
@davidtaylorhq thanks for the excellent feedback! I believe I was able to implement most of it or at least the general idea.

One thing I didn't realize is that the slug is dynamic and ends up being an unreliable way to find the text customization key as soon as the badge name is customized 😢 

I'm not sure if you'll be excited about what I've done in the controller, so I totally understand if that needs work. I racked my brain for a good chunk of the day trying to come up with a solution. If what I came up is too hacky to be acceptable, let me know if you have any ideas. Otherwise I may have to put this to rest.

**Comment by davidtaylorhq** [comment]
This is looking good 👍 

I see what you mean about the controller logic - and yeah we really want to avoid an extra AJAX request if possible. I was looking at how we handle this in the backend. The `Badge` model has an `i18n_name` method

https://github.com/discourse/discourse/blob/97eae4e3add4c31841ddc4858dc62c68ef86b1e6/app/models/badge.rb#L298-L300

If we move that above the `protected` section in the model, then we can add it to the admin badge serializer here:

https://github.com/discourse/discourse/blob/97eae4e3add4c31841ddc4858dc62c68ef86b1e6/app/serializers/admin_badge_serializer.rb#L4-L4

^^ This serializer is only used in the admin panel, so the change will have no effect on regular users at all. We could add an `include_i18n_name?` method so that it only gets serialized for system badges.

And then, that i18n_name value should be available in the js/hbs for use.

**Comment by tshenry** [comment]
Sorry for the radio silence here. I had a good opportunity work on this at the end of the day today. Your suggested solution works beautifully! Thanks for outlining what needed to be done in such detail 😍 

I'm out of time today, but do you think this should include some kind of test before this is merged? I so, do you think I should go with a unit test for the link presence/functionality? Should there be a spec for `include_i18n_name?`, or is that simple enough?

File: app/assets/javascripts/admin/addon/controllers/admin-badges-show.js
Comment (davidtaylorhq) [inline app/assets/javascripts/admin/addon/controllers/admin-badges-show.js:9]: Not sure what happened here? Looks like thee imports got reordered when compared to core? Would be good to try and remove this diff.

**Comment by davidtaylorhq** [comment]
This all looks good to me. For a spec, you could add something to `spec/requests/admin/badges_controller_spec` to check that the i18n name is being serialized for system badges.

I don't think we currently have any JS tests for the admin badges UI, which isn't ideal... If you wanted to try and add some that would be great, but I don't think it's absolutely essential for this PR to be merged. The change here is pretty low risk.

**Review by e1r2-2864 — APPROVED** [review]

**Comment by SamSaffron** [comment]
@tshenry  we are good to merge here once you correct the import order David mentioned.

**Review by davidtaylorhq — APPROVED** [review]
❤️

**Pull request merged**
